pkg/util/clienthelper/clienthelper.go

```diff
@@ -0,0 +1,273 @@
+	if err != nil {
+		return nil, err
+	}
+	return NewWithClient(log, client), nil
+}
+
+func NewWithClient(log *logrus.Entry, client client.Client) Interface {
+	return &dynamicHelper{
+		log:    log,
+		client: client,
+	}
+}
+
+func (dh *dynamicHelper) EnsureDeleted(ctx context.Context, gvk schema.GroupVersionKind, key types.NamespacedName) error {
+	a := meta.AsPartialObjectMetadata(&metav1.ObjectMeta{
+		Name:      key.Name,
+		Namespace: key.Namespace,
+	})
+	a.SetGroupVersionKind(gvk)
+
+	dh.log.Infof("Delete kind %s ns %s name %s", gvk.Kind, key.Namespace, key.Name)
+	err := dh.client.Delete(ctx, a)
+	if kerrors.IsNotFound(err) {
+		return nil
+	}
+	return err
+}
+
+func (dh *dynamicHelper) GetOne(ctx context.Context, key types.NamespacedName, obj kruntime.Object) error {
+	newObj, ok := obj.(client.Object)
+	if !ok {
+		return errors.New("can't convert object")
+	}
+
+	return dh.client.Get(ctx, key, newObj)
+}
+
+// Ensure that one or more objects match their desired state.  Only update
+// objects that need to be updated.
+func (dh *dynamicHelper) Ensure(ctx context.Context, objs ...kruntime.Object) error {
+	for _, o := range objs {
+		err := dh.ensureOne(ctx, o)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (dh *dynamicHelper) ensureOne(ctx context.Context, new kruntime.Object) error {
+	gvk, err := apiutil.GVKForObject(new, scheme.Scheme)
+	if err != nil {
+		return err
+	}
+
+	newObj, ok := new.(client.Object)
+	if !ok {
+		return fmt.Errorf("object of kind %s can't be made a client.Object", gvk.String())
+	}
+
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		old, err := scheme.Scheme.New(gvk)
+		if err != nil {
+			return err
+		}
+
+		oldObj, ok := old.(client.Object)
+		if !ok {
+			return fmt.Errorf("object of kind %s can't be made a client.Object", gvk.String())
+		}
+
+		err = dh.client.Get(ctx, client.ObjectKey{Namespace: newObj.GetNamespace(), Name: newObj.GetName()}, oldObj)
+		if kerrors.IsNotFound(err) {
+			dh.log.Infof("Create %s", keyFunc(gvk.GroupKind(), newObj.GetNamespace(), newObj.GetName()))
+			return dh.client.Create(ctx, newObj)
+		}
+		if err != nil {
+			return err
+		}
+		candidate, changed, diff, err := merge(oldObj, newObj)
+		if err != nil || !changed {
+			return err
+		}
+		dh.log.Infof("Update %s: %s", keyFunc(gvk.GroupKind(), candidate.GetNamespace(), candidate.GetName()), diff)
+		return dh.client.Update(ctx, candidate)
+	})
+}
+
+// merge takes the existing (old) and desired (new) objects.  It compares them
+// to see if an update is necessary, fixes up the new object if needed, and
+// returns the difference for debugging purposes.
+func merge(old, new client.Object) (client.Object, bool, string, error) {
+	if reflect.TypeOf(old) != reflect.TypeOf(new) {
+		return nil, false, "", fmt.Errorf("types differ: %T %T", old, new)
+	}
+
+	// 1. Set defaults on new.  This gets rid of many false positive diffs.
```

Comment:
I think the explanation on this needs to be a little more explicit. I think what you're saying is that it does things like strips away "creationDate" etc. ?